File: temporal-tutorial/temporal_worker/workflows.py
from datetime import timedelta
from temporalio import workflow
from temporalio.common import RetryPolicy
from temporalio.exceptions import ApplicationError
from models import PaymentDetails, WithdrawRequest, DepositRequest
from activities import BankingActivities

import logging
logger = logging.getLogger("temporal_worker.product.workflows")

# Common retry policy for banking activities
COMMON_RETRY_POLICY = RetryPolicy(
    initial_interval=timedelta(seconds=5),
    maximum_attempts=10,
    backoff_coefficient=2.0,
    maximum_interval=timedelta(seconds=10)
)

@workflow.defn
class MoneyTransferWorkflow:
    """Baseline money transfer workflow - 2 step process"""
    
    @workflow.run
    async def run(self, payment_details: PaymentDetails) -> str:
        """Execute money transfer workflow"""
        logger.info(f"Starting money transfer: {payment_details.from_account} -> {payment_details.to_account}, amount: {payment_details.amount}")
        
        # Step 1: Withdraw money from source account
        withdraw_request = WithdrawRequest(
            account_id=payment_details.from_account,
            amount=payment_details.amount,
            reference=payment_details.reference
        )
        
        withdraw_result = await workflow.execute_activity(
            BankingActivities.withdraw,
            withdraw_request,
            start_to_close_timeout=timedelta(seconds=30),
            retry_policy=COMMON_RETRY_POLICY
        )
        
        logger.info(f"Withdrawal successful: {withdraw_result.transaction_id}")
        
        # Step 2: Deposit money to destination account
        deposit_request = DepositRequest(
            account_id=payment_details.to_account,
            amount=payment_details.amount,
            reference=payment_details.reference
        )
        
        deposit_result = await workflow.execute_activity(
            BankingActivities.deposit,
            deposit_request,
            start_to_close_timeout=timedelta(seconds=30),
            retry_policy=COMMON_RETRY_POLICY
        )
        
        logger.info(f"Deposit successful: {deposit_result.transaction_id}")
        logger.info(f"Money transfer completed successfully")
        
        return f"Transfer complete: {withdraw_result.transaction_id} -> {deposit_result.transaction_id}"

File: temporal-tutorial/ts_worker/src/workflows.ts
// Application workflows: pure business logic. Signadot routing is handled by
// platform-owned interceptors (see src/signadot) wired up by SandboxAwareWorker.
import { proxyActivities, log } from '@temporalio/workflow';
import type * as activities from './activities';
import type { PaymentDetails } from './models';

// Common retry policy for banking activities
const { withdraw, deposit } = proxyActivities<typeof activities>({
  startToCloseTimeout: '30 seconds',
  retry: {
    initialInterval: '5 seconds',
    maximumAttempts: 10,
    backoffCoefficient: 2.0,
    maximumInterval: '10 seconds',
  },
});

/** Baseline money transfer workflow - 2 step process */
export async function MoneyTransferWorkflow(paymentDetails: PaymentDetails): Promise<string> {
  log.info(
    `Starting money transfer: ${paymentDetails.from_account} -> ${paymentDetails.to_account}, amount: ${paymentDetails.amount}`
  );

  // Step 1: Withdraw money from source account
  const withdrawResult = await withdraw({
    account_id: paymentDetails.from_account,
    amount: paymentDetails.amount,
    reference: paymentDetails.reference,
  });
  log.info(`Withdrawal successful: ${withdrawResult.transaction_id}`);

  // Step 2: Deposit money to destination account
  const depositResult = await deposit({
    account_id: paymentDetails.to_account,
    amount: paymentDetails.amount,
    reference: paymentDetails.reference,
  });
  log.info(`Deposit successful: ${depositResult.transaction_id}`);

  log.info('Money transfer completed successfully');
  return `Transfer complete: ${withdrawResult.transaction_id} -> ${depositResult.transaction_id}`;
}
